backend/tracim_backend/applications/agenda/application.py

```diff
@@ -0,0 +1,114 @@
+        """
+        Check if config is correctly setted for caldav features
+        """
+        app_config.check_mandatory_param(
+            "CALDAV__RADICALE_PROXY__BASE_URL",
+            app_config.CALDAV__RADICALE_PROXY__BASE_URL,
+            when_str="when caldav feature is enabled",
+        )
+        # TODO - G.M - 2019-05-06 - convert "caldav.radicale.storage.filesystem_folder"
+        # as tracim global parameter
+        app_config.check_mandatory_param(
+            "CALDAV__RADICALE__STORAGE__FILESYSTEM_FOLDER",
+            app_config.CALDAV__RADICALE__STORAGE__FILESYSTEM_FOLDER,
+            when_str="if caldav feature is enabled",
+        )
+        app_config.check_directory_path_param(
+            "CALDAV__RADICALE__STORAGE__FILESYSTEM_FOLDER",
+            app_config.CALDAV__RADICALE__STORAGE__FILESYSTEM_FOLDER,
+            writable=True,
+        )
+        radicale_storage_type = app_config.settings.get("caldav.radicale.storage.type")
+        if radicale_storage_type != "multifilesystem":
+            raise ConfigurationError(
+                '"{}" should be set to "{}"'
+                " (currently only valid value)"
+                ' when "{}" is true'.format(
+                    "caldav.radicale.storage.type", "multifilesystem", "caldav.enabled"
+                )
+            )
+
+    def import_controllers(
+        self,
+        configurator: Configurator,
+        app_config: CFG,
+        route_prefix: str,
+        context: PyramidContext,
+    ) -> None:
+        # TODO - G.M - 2019-03-18 - check if possible to avoid this import here,
+        # import is here because import AgendaController without adding it to
+        # pyramid make trouble in hapic which try to get view related
+        # to controller but failed.
+        from tracim_backend.applications.agenda.controllers.agenda_controller import (
+            AgendaController,
+        )
+        from tracim_backend.applications.agenda.controllers.radicale_proxy_controller import (
+            RadicaleProxyController,
+        )
+
+        configurator.include(add_www_authenticate_header_for_caldav)
+        # caldav exception
+        context.handle_exception(CaldavNotAuthorized, HTTPStatus.FORBIDDEN)
+        context.handle_exception(CaldavNotAuthenticated, HTTPStatus.UNAUTHORIZED)
+        # controller
+        radicale_proxy_controller = RadicaleProxyController(
+            proxy_base_address=app_config.CALDAV__RADICALE_PROXY__BASE_URL,
+            radicale_base_path=app_config.CALDAV__RADICALE__BASE_PATH,
+            radicale_user_path=app_config.CALDAV__RADICALE__USER_PATH,
+            radicale_workspace_path=app_config.CALDAV_RADICALE_WORKSPACE_PATH,
+        )
+        agenda_controller = AgendaController()
+        configurator.include(agenda_controller.bind, route_prefix=BASE_API_V2)
+        configurator.include(radicale_proxy_controller.bind)
+
+
+def create_app() -> TracimApplication:
```

Comment:
To import app, we do need to know where to find them. After discussion with @buxx about multiple solution, we decide to have a "create_app" method in application.py module of each applicaiton folder.

- Convention naming instead of automatic subclass fetching because subclass iterating, subclass iterating force to do recursive iteration and is less clear/more magic.
- Only this method outside of class here, allowing relying on explicit documentation for other methods with subclassing.
- Method instead of application object to avoid staticness (this may cause trouble during running test)

Reply:
`create_app()` is not a *method* but a *function* - it is important to use the right vocabulary to make discussion clear. I do not see why a class method would be more static than function, could you explain?

Reply:
Problem is about naming. If we want to import class or function, we do need to know his name.
Same name class is a bad idea for debugging, so function seems a better solution.

Reply:
>  I do not see why a class method would be more static than function, could you explain?

@lebouquetin 

Other solution was not a class method but a variable named application (like in wsgi files), like this:

**backend/tracim_backend/applications/example/application.py**

```python
# [...]

class ExampleApp(TracimApplication):
    # [...]

application = ExampleApp()
```

Bad side of this solution was:

*  ̀ExampleApp` instantiate at module loading
* `application` is static